reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+                                    pullIntoDescriptor.byteLength - pullIntoDescriptor.bytesFilled]);
+  } catch (e) {
+    DestroyReadableByteStreamController(controller);
+    const stream = controller._controlledReadableByteStream;
+    if (stream._state === 'readable') {
+      ErrorReadableByteStream(stream, e);
+    }
+  }
+
+  controller._insideUnderlyingByteSource = false;
+}
+
+function CallPullOrPullIntoLaterIfNeeded(controller) {
+  if (controller._insideUnderlyingByteSource) {
+    controller._callPullOrPullIntoLaterIfNeeded = true;
+    return;
+  }
+
+  Promise.resolve().then(CallPullOrPullIntoRepeatedlyIfNeeded.bind(undefined, controller));
+}
+
+function CallPullOrPullIntoRepeatedlyIfNeeded(controller) {
+  const stream = controller._controlledReadableByteStream;
+
+  while (true) {
+    if (!controller._callPullOrPullIntoLaterIfNeeded) {
+      return;
+    }
+
+    if (controller._closeRequested) {
+      return;
+    }
+    if (stream._state !== 'readable') {
+      return;
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined) {
+      return;
+    }
+
+    if (IsReadableByteStreamReader(reader)) {
+      if (reader._readRequests.length === 0) {
+        return;
+      }
+
+      CallPull(controller);
+    } else {
+      assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+      if (reader._readIntoRequests.length === 0) {
+        return;
+      }
+
+      CallPullInto(controller);
+    }
+  }
+}
+
+function CancelReadableByteStream(stream, reason) {
+  if (stream._state === 'closed') {
+    return Promise.resolve(undefined);
+  }
+  if (stream._state === 'errored') {
+    return Promise.reject(stream._storedError);
+  }
+
+  CloseReadableByteStream(stream);
+
+  const controller = stream._controller;
+
+  controller._totalQueuedBytes = 0;
+  controller._queue = [];
+
+  const sourceCancelPromise = PromiseInvokeOrNoop(controller._underlyingByteSource, 'cancel', [reason]);
+  return sourceCancelPromise.then(() => undefined);
+}
+
+function CloseReadableByteStream(stream) {
+  assert(IsReadableByteStream(stream), 'stream must be ReadableByteStream');
+  assert(stream._state === 'readable', 'state must be readable');
+
+  stream._state = 'closed';
+
+  if (IsReadableByteStreamLocked(stream)) {
+    CloseReadableByteStreamReader(stream._reader);
+  }
+}
+
+function CloseReadableByteStreamReader(reader) {
+  reader._state = 'closed';
+
+  reader._closedPromise_resolve(undefined);
+  reader._closedPromise_resolve = undefined;
+  reader._closedPromise_reject = undefined;
+
+  if (IsReadableByteStreamReader(reader)) {
+    for (const req of reader._readRequests) {
+      req.resolve(CreateIterResultObject(undefined, true));
+    }
+
+    reader._readRequests = [];
+  } else {
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    if (reader._readIntoRequests.length > 0) {
+      return;
```

Comment:
This needs a comment or something ... when you close a BYOB reader that has readInto requests, it doesn't get detached? Would it be better to not close it yet, and e.g. set closeRequested or something?

Apologies since this is probably something we discussed previously. It is complicated >_<